### PR TITLE
Added ViewRenderer

### PR DESCRIPTION
**Pull request by bre7**
Closes https://github.com/vapor/vapor/issues/1637

### CODE REVIEW

File: Sources/TemplateKit/Utilities/ViewRenderer.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import Foundation
```
Comment (tanner0101) [inline Sources/TemplateKit/Utilities/ViewRenderer.swift]: this file should go in the same folder that `View` is defined

File: Sources/TemplateKit/Utilities/ViewRenderer.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+/// View renderers are responsible for loading
+/// the files paths given and caching if needed.
+public protocol ViewRenderer: class {
+
+    /// For view renderers that use a cache to
```
Comment (tanner0101) [inline Sources/TemplateKit/Utilities/ViewRenderer.swift]: old comment style from vapor 2, these should be widened up.

**Review by tanner0101 — COMMENTED** [review]
There should also be a free extension conforming template renderers to this protocol. That way we just need to add this protocol to LeafRenderer and everything will work automatically.

```swift
extension ViewRenderer where Self: TemplateRenderer {
   /// implement all of the methods
}
```

File: Sources/TemplateKit/Utilities/ViewRenderer.swift (hunk 2 above)
Comment (bre7) [reply Sources/TemplateKit/Utilities/ViewRenderer.swift]: How many char per line ? (todo: Add `.swiftlint`)

Comment (tanner0101) [reply Sources/TemplateKit/Utilities/ViewRenderer.swift]: uh.... just wider haha. there should probably be a clearer standard >_<. Take a look at some of the other files to see

File: Sources/TemplateKit/ViewRenderer.swift [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+/// View renderers are responsible for loading the files paths given and caching if needed.
+///
+/// View renderers are also responsible for accepting a Node for templated responses.
```
Comment (tanner0101) [inline Sources/TemplateKit/ViewRenderer.swift]: oh man,`Node`. That is an old doc block.

Comment (tanner0101) [reply Sources/TemplateKit/ViewRenderer.swift]: This should mention codable instead.

Comment (bre7) [reply Sources/TemplateKit/ViewRenderer.swift]: Use the same as TemplateRenderer: `/// Renders raw template data (bytes) to `View`s.` ?

Comment (tanner0101) [reply Sources/TemplateKit/ViewRenderer.swift]: maybe:

> `ViewRenderer` returns a `View` for a supplied path an `Encodable` object.

ViewRenderer doesn't have to work directly with raw bytes at all, for instance with mocked versions

File: Sources/TemplateKit/ViewRenderer.swift [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    /// - parameters:
+    ///     - path: Path to file contianing raw template bytes.
+    ///     - context: `Encodable` item that will be encoded to `TemplateData` and used as template context.
+    /// - returns: `Future` containing the rendered `View`.
+    func render<E>(_ path: String, _ context: E) -> Future<View> where E: Encodable
+}
+
+extension ViewRenderer where Self: TemplateRenderer {
+
+    public func render<E>(_ path: String, _ context: E) -> Future<View> where E: Encodable {
```
Comment (bre7) [inline Sources/TemplateKit/ViewRenderer.swift]: Avoid code dupe using `return (self as TemplateRenderer).render(...)` ?

File: Sources/TemplateKit/ViewRenderer.swift [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    ///
+    /// - parameters:
+    ///     - path: Path to file contianing raw template bytes.
+    ///     - context: `Encodable` item that will be encoded to `TemplateData` and used as template context.
+    /// - returns: `Future` containing the rendered `View`.
+    func render<E>(_ path: String, _ context: E) -> Future<View> where E: Encodable
+}
+
+extension ViewRenderer where Self: TemplateRenderer {
+
```
Comment (tanner0101) [inline Sources/TemplateKit/ViewRenderer.swift]: We should have an extension for `shouldCache`, too.

Comment (bre7) [reply Sources/TemplateKit/ViewRenderer.swift]: `true` by default ?

File: Sources/TemplateKit/ViewRenderer.swift [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    public var shouldCache: Bool { return true }
```
Comment (tanner0101) [inline Sources/TemplateKit/ViewRenderer.swift:23]: Should cache should be implemented by setting [`astCache`](https://github.com/vapor/template-kit/blob/master/Sources/TemplateKit/Pipeline/TemplateRenderer.swift#L24) to `nil` when false and initializing it to an empty cache when true.

**Comment by tanner0101** [comment]
Once DatabaseKit is done being tagged, I will be working on Vapor and i'll be sure to grab this one 👍

**Comment by bre7** [comment]
Great !

**Pull request merged**